process/datadog_checks/process/process.py

```diff
@@ -109,63 +215,72 @@ def find_pids(self, name, search_string, exact_match, ignore_ad=True):
 
         refresh_ad_cache = self.should_refresh_ad_cache(name)
 
-        encountered_process_names = set()
         matching_pids = set()
 
-        for proc in psutil.process_iter():
-            # Skip access denied processes
-            if not refresh_ad_cache and proc.pid in self.ad_cache:
-                continue
+        # Acquire the write lock to check whether to refresh because we're
+        # going to keep it to do the refresh, AND, we don't want multiple
+        # threads getting a `yes` result at once
+        with self.process_list_cache.write_lock():
+            if self.should_refresh_proclist():
+                self.log.debug("Refreshing process list")
+                self.process_list_cache.elements = [proc for proc in psutil.process_iter(attrs=['pid', 'name'])]
```

Comment:
I don't think it works unfortunately. You're creating a element list local to this `process_list_cache` instance.

I think the best fix is to make `ProcessListCache` a class attribute, so that you only have one instance.

It'd be nice to have a test before that though, making sure we don't call psutil uselessly.